Add the reset stragegy for checking out objects

TyOverby opened the pull request:
This pull request changes the name of the `checkout` function to `checkout_raw` and adds a new function named `checkout` for objects that can be "reset" back to a neutral state.

carllerche commented:
I do think that being able to reset the value before checkout is valuable, but I don't know it is the common case. Renaming the current `checkout()` function -> `checkout_raw() may not be the best strategy. Perhaps there should be a decorator around`Pool`that calls`reset()` on checkout?

TyOverby commented:
What do you mean by decorator? 

What if we had `checkout_dirty` and `checkout_cleaned`?  Neither sound "default", but are pretty clear about what is going on.

carllerche commented:
Something like:

``` rust
impl<T> Clean<Pool<T>> {
    pub fn checkout(&mut self) -> Checkout<T> {
        self.pool.checkout().reset
    }
}
```

Just an idea of course.

My main thought is that the behavior  you would want depends on T. So, for a given T, you would always want the value reset on checkout or you would never want it. If this is the case, it would make more sense to have a separate pools that have different behavior on checkout.

What do you think?

TyOverby commented:
I really agree with the commentor on reddit that said that returning "dirty" objects is not the expected behavior.  By having a wrapper type (`Clean`), it makes it look like Clean pools are a special case.

carllerche commented:
I think that I am coming around. A few requests and I think I will be ok w/ this PR

Instead of having `checkout_raw`, why not provide a `Dirty<T>` type that is a simple wrapper that also implements `Deref`, `DerefMut`, `Sync` (when `T` is `Sync`), `Send` (when `T` is `Send), and implement`Reset`on it to be a no-op. This way, one can have the equivalent of`checkout_raw`by doing:`Pool<Dirty<MyType>>`.

Update the docs to have some notes on resetting & dirty?

TyOverby commented:
Cool.  I'll add those changes tonight.  I think that's a good API!

TyOverby commented:
Alright, I added a `Dirty` struct and implemented Reset for it.  I also implement Reset for all objects that implement `Default` and `Clone`, which hits a large amount of objects, and with the `clone_from` method, you can [keep the capacity](http://is.gd/ZHFLiE) given to you by many rust collections.

Travis CI is failing because because your doc comment uses `thread::scoped`.

TyOverby commented:
@carllerche: what do you think of the current state?

carllerche commented on src/reset.rs:
```diff
@@ -0,0 +1,40 @@
+
+// For most of the stdlib collections, this will "clear" the collection
+// without deallocating.
+impl <T: Default + Clone> Reset for T {
```
I don't think that we can include this blanket implementation or it will prevent consumers of the library from being able to implement `Reset` on their types.

carllerche commented:
Sorry for the delay. I was at Railsconf. I left a note on a blanket implementation that should be removed, but besides that, it looks good to me. Could you also rebase on master? That should get CI to pass. Thanks!

TyOverby commented:
I think that having a really easy way to clean objects is important, so I added a new struct like `Dirty` called `Default` that uses the `Default + Clone` hack to reset objects.

Let me know if you think this is ok, otherwise I'll drop it and just keep the `Dirty` struct.

carllerche commented:
I believe that I was wrong re: a blanket implementation preventing implementations by consumers. Sorry about that (I'm still fuzzy on coherence rules). I have merged the commit before introducing `Default`. c929ed480632f0ecdd88d9aa459b21487b2382fc. I'm going to close this for now, but if there are still changes required, feel free to open up a new ticket / PR. Thanks for the work :heart:
